### PR TITLE
highlighting: core

### PR DESCRIPTION
**Pull request by PLNech**

### CODE REVIEW

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+) {
+
+    private val regex = Regex(Regex.escape(preTag) + ".*?" + Regex.escape(postTag))
+
+    fun tokenize(input: String): HighlightedString {
+        return HighlightedString(input, mutableListOf<HighlightToken>().apply {
+            var startIndex = 0
+            var matchResult = regex.find(input)
+
+            while (matchResult != null && matchResult.groups.isNotEmpty()) {
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt]: If you use `regex.findAll`, probably we can do it without a while loop

File: core/src/commonTest/kotlin/highlighting/TestHighlightTokenizer.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
+package highlighting
```
Comment (q-litzler) [inline core/src/commonTest/kotlin/highlighting/TestHighlightTokenizer.kt]: I would also test `matchInTheMiddle`, `matchInTheEnd`, `matchInTheStart`

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightToken.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.algolia.instantsearch.core.highlighting
+
+
+/**
+ * A token from a highlighted string.
+ */
+public data class HighlightToken(val highlighted: Boolean = false, val content: String)
```
[the remark this answers is not in the document]
Comment (PLNech) [reply core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightToken.kt]: Agreed, surprised I missed that!

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package com.algolia.instantsearch.core.highlighting
+
+
+private const val DEFAULT_PRETAG: String = "<em>"
```
[the remark this answers is not in the document]
Comment (PLNech) [reply core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt]: I'm reusing it later, but still you're right on naming

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+private const val DEFAULT_POSTTAG: String = "</em>"
+
+public class HighlightTokenizer(
+    public val preTag: String = DEFAULT_PRETAG,
+    public val postTag: String = DEFAULT_POSTTAG
+) {
+
+    private val regex = Regex(Regex.escape(preTag) + ".*?" + Regex.escape(postTag))
+
+    fun tokenize(input: String): HighlightedString {
```
[the remark this answers is not in the document]
Comment (PLNech) [reply core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt]: Good call

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt (hunk 1 above)
Comment (PLNech) [reply core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt]: I started with `findAll` and it didn't work out well, but I'll give it another try!

File: core/src/commonTest/kotlin/highlighting/TestHighlightTokenizer.kt (hunk 2 above)
Comment (PLNech) [reply core/src/commonTest/kotlin/highlighting/TestHighlightTokenizer.kt]: Sure!

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package com.algolia.instantsearch.core.highlighting
+
+
+private const val DefaultPreTag: String = "<em>"
+private const val DefaultPostTag: String = "</em>"
+
+public class HighlightTokenizer(
+    public val preTag: String = DefaultPreTag,
+    public val postTag: String = DefaultPostTag
+) : ((String) -> HighlightedString) {
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt]: Extraneous parenthesis

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightedString.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.algolia.instantsearch.core.highlighting
+
+
+data class HighlightedString(val original: String, val parts: List<HighlightToken>) {
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightedString.kt]: public

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt (hunk 6 above)
Comment (PLNech) [reply core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightTokenizer.kt]: Right

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightedString.kt (hunk 7 above)
Comment (PLNech) [reply core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightedString.kt]: Indeed

**Pull request merged**